### ISSUE TITLE
zellij frequent disk write activity

### ISSUE THREAD

**Issue by fungos** [zellij-org/zellij#3127]
**Basic information**

`zellij --version`: 0.39.2

`stty size`:

`uname -av` or `ver`(Windows): Linux x 6.6.10-1-MANJARO #1 SMP PREEMPT_DYNAMIC x86_64 GNU/Linux

## Further information
Reproduction steps, noticeable behavior, related issues, etc

# 2. Issues with the Zellij UI / behavior / crash

## Issue description

While running in foreground or even detached, the disk activity of zellij seems to do too frequent writes on disk. I have an unusual noisy disk, and I detect what I consider bad behavior on software by the noise the disk does. After trying zellij for a few days, I noted my disk was a lot more noisy than habitual. Looking at `sudo iotop` I can see zellij doing lots of tiny writes too frequent which is causing my disk to spin when my computer should be idle. Other software causes this kind of writes, mostly writing sessions to disk, like firefox. But the frequency of the writes is not that elevated which keeps the noise to somewhat acceptable threshold, but zellij in comparison to these other seem to be doing something between 3-5 more writes.

**Comment by pedromfedricci** [comment]
Zellij takes snapshots of all active sessions every 1sec ([link](https://zellij.dev/documentation/session-resurrection#session-files-in-the-cache)). I guess that _might_ be all the noise that you are seeing. Looking at the docs, there is no way to change the default interval, but you can disable it all together with `session_serialization false` ([link](https://zellij.dev/documentation/options#session_serialization)) in your config. Try running Zellij with the serialization option disabled to see if that is the issue.
Another option is that there is indeed some abnormal behavior going on. Figuring it out the root cause will require some more digging. Monitor Zellij's log at `/tmp/zellij-1000/zellij-log/zellij.log` while running it, look out for patterns that could hint what is the problem. It is even possible that the frequent writes you are seeing are actually Zellij writing to the log file constantly.

**Comment by fungos** [comment]
I can confirm it is the session serialized, maybe exposing a setting to change the frequence would be nice to have.

**Comment by shanesveller** [comment]
Configuration option sounds like a welcome step forward - for example I personally would dial it down to something more like 60sec-5min on my own systems given the opportunity. If it doesn't already, maybe also ensure that the session-quit command (Ctrl-O Q?) does one last flush of the current state to the serialized copy.

**Comment by sandorex** [comment]
I believe this is already an option `serilization_interval` at https://github.com/zellij-org/zellij/blob/b677ffe75fb8e518441a0bd7df02abfb8dcc4989/zellij-utils/src/input/options.rs#L153C9-L153C31
But the caveat is that its a commandline argument not parsed from config from what i can see

**Comment by cristiand391** [comment]
From the code and this comment it seems it's available as a config var: https://github.com/zellij-org/zellij/issues/2946#issuecomment-1814436010

I've seen some other options undocumented (they aren't generated from code).

**Comment by cristiand391** [comment]
Also, in this PR the default serialization interval changed from 1s to 60s:
https://github.com/zellij-org/zellij/pull/2923

**Comment by pedromfedricci** [comment]
PR #2923 (v0.39.1) introduced the serialization interval config (not documented yet) and PR #2951 (v0.39.2) changed the default interval from 1sec to 60sec for how often the layout is **reported** as ready to be written into disc. The background job still writes into both **session-metadata.kdl** and **session-layout.kdl** (only after its first report in this case) files every second for all active sessions ([code](https://github.com/zellij-org/zellij/blob/27bffbf1533b4b2d3c10b1305557c75ddd121374/zellij-server/src/background_jobs.rs#L178)), regardless of what the `serialization_interval` value is. My reading so far is that this is indeed a bug.

**Comment by justinlovinger** [comment]
I disabled `session_serialization`, but I still see a consistent 4KB/s disk-write activity for every session, alongside constant CPU-usage, even when idle. I am on version 0.39.2.

**Comment by pedohorse** [comment]
> 4KB/s disk-write activity

For me it's 150-200 KB/s of read and around 80-100 KB/s of write. I have 3 sessions almost always running, each reads the whole config each second, and writes seemingly only it's part to it.

So far I was just hoping that kernel/fs-driver optimizes those writes somehow and it's not actually always writes to the device.

It would be very nice to have that thing configurable.

**Comment by imsnif** [comment]
In addition to [disabling session serialization](https://zellij.dev/documentation/options#session_serialization) you can now also disable writing the session metadata to disk with: `disable_session_metadata true/false` - note that this might cause features such as the session-manager not to work properly.

**Comment by pedohorse** [comment]
I've made a little draft PR just to illustrate one possible solution to the issue - just making the poll interval configurable

#3298

**Comment by pedohorse** [comment]
there is `serialization_interval`, but that PR is about a different polling interval.
Even when not serializing anything, zellij is loading all session data, checks it and saves it back.
this is how one instance of zellij can always see all session info in the session manager.

the downside of increasing this interval is that you will not see up-to-date session information in the session manager

**Comment by vi** [comment]
Maybe it should save session information only when something is changed (i.e. still run a timer, but skip serializing when nothing has changed) and load all session data only when that session manager is being accessed?

**Comment by pedohorse** [comment]
> Maybe it should save session information only when something is changed (i.e. still run a timer, but skip serializing when nothing has changed) and load all session data only when that session manager is being accessed?

i can add my opinion since noone else answered yet:
(disclaimer: I've looked into the code once so this is my surface opinion) 
it always reads/writes the session file because it uses that file to sync session managers that are running on different zellij instances. So there is no way to know that session file was changed unless you constantly poll it (or rely on file metadata?).
But it does feel that writing it back is unnecessary if nothing has changed in the current session

**Comment by Dietr1ch** [comment]
>  it always reads/writes the session file because it uses that file to sync session managers that are running on different zellij instances. So there is no way to know that session file was changed unless you constantly poll it (or rely on file metadata?).

Synchronising multiple instances on a file seems alright, but only as long as that file is in-memory.

I feel that writes to persistent disk are only necessary on events like shutting down or explicit user saves, but permissible under timers that let disks idle/sleep, so at least 2 minutes, preferably 5+


>  But it does feel that writing it back is unnecessary if nothing has changed in the current session

This would probably help unless something silly like a timestamp is sneaking in and changing the file every time.

**Comment by pftbest** [comment]
I feel like this should be more visible. In the modern world of SSD drives with limited lifespan, doing a lot of small writes feels like a waste. Specifying both options in the config file fixed the issue for me.
```
session_serialization false
disable_session_metadata true
```

EDIT: Just to clarify the magnitude of the issue, in case someone is not convinced

<img width="923" alt="Image" src="https://github.com/user-attachments/assets/74dbe48b-d6da-474c-8518-aaa899f245e8" />

**Comment by permahorse** [comment]
yet it would be so nice to both have session serialization, and not have constant disk activity... :)

also, as I was watching my SSD status over last year since noticing this issue - it **seeeeeems** that OS/firmware is optimizing these small writes, so actual impact on the hardware **seem** to be negligible.

However I still very much want to get rid of those constant disk writes

**Comment by SheliakLyr** [comment]
Hi, I would like to raise awareness of this issue. I also suggest changing the defaults to disable session serialization until a better solution is found.

I tried to use zellij, but after a few hours my system became completely unusable. A have multiple SSDs, and I suspected that one of them may be failing for various reasons. I saw massive iowait spikes that could freeze my UI for a few seconds. This was happening during heavy I/O operations (very large gradle builds), so initially I did not suspect zellij at al.

However, from iotop I could see a lot of writes from zellij processes (50-200kB/s). Even when those sessions were idle, I/O traffic was constant. So I have decided to kill all zellij sessions and problems disappeared immediatelly.

After session/metadata serialization is disabled, zellij can be used without any noticable impact on the system.

It is possible, that other operations on the system enlarge this issue... or maybe the cause is in the SSD. But I believe that zellij default approach to session serialization is the root cause.

**Comment by Dietr1ch** [comment]
I snooped at zellij with `lsof` and it says it's using these directories,

- `$XDG_CACHE_HOME/zellij/`
- `$XDG_RUNTIME_DIR/zellij/`
- `/tmp/zellij-$UID/` (why not $XDG_RUNTIME_DIR?)

My config is set to try to avoid some writes.

```kdl
# ~/.config/zellij/config.kdl
disable_session_metadata true
serialize_pane_viewport false
session_serialization false
```

I'm not sure if this is enough today. I feel we need a global option to avoid writes, as new write paths would probably by default work around a config that disables known write paths one by one.



On my system `$XDG_RUNTIME_DIR` and `/tmp` are on memory, so writing there would be fine, but I'm not sure of where the recurrent writes are really going to.

**Comment by altunyurt** [comment]
A workaround until a permanent solution is found is to use /dev/shm for the Zellij cache, sacrificing session persistence to save the disk from frequent I/O. On my setup, I don't need sessions to survive a reboot or a crash, so this solution works perfectly for my needs.

```conf
#    ~/.config/user-tmpfiles.d/ram-caches.conf
# 'd' creates the path if it does not exist. 
# 'L+' creates the symlink, overwriting anything that currently exists at the destination path.

# Type  Path                     Mode  User  Group  Age  Argument (Target)
d       /dev/shm/cache/zellij    0700  -     -      -    -
L+      %h/.cache/zellij         -     -     -      -    /dev/shm/cache/zellij
d       /dev/shm/cache/mozilla    0700  -     -      -    -
L+      %h/.cache/mozilla        -     -     -      -    /dev/shm/cache/mozilla
#  ...
```

Save the configuration file and run the following command:

```sh
$ systemd-tmpfiles --user --create
```

You'll need to make sure ```systemd-tmpfiles-setup.service``` is enabled for the user or globally, so that the user-tmpfiles config read on each reboot or login .
